src/programs/normalisefasta.1
.TH NORMALISEFASTA 1 "January 2014" BIOBAMBAM
.SH NAME
normalisefasta - normalise line length in a FastA file
.SH SYNOPSIS
.PP
.B normalisefasta
[options]
.SH DESCRIPTION
normalisefasta reads a FastA file from standard input and outputs a
reformatted version of the file with a consistent length of the lines
containing sequence information on standard output. The program can either
produce uncompressed or BGZF compressed output. For uncompressed output
a FastA index (.fai) is produced on the standard error channel.
.PP
The following key=value pairs can be given:
.PP
.B cols=<[80]>
line width for the lines containing sequence information in number of bases.
This option is only considered for uncompressed output (i.e. bgzf=0)
.PP
.B bgzf=<0|1>
produce uncompressed (bgzf=0) or compressed (bgzf=1) output
.PP
.B index=<>
if bgzf=1 this key can be used for giving the file name for the index file
allowing (pseudo) random access in the output file. If the key is not given
when bgzf=1, then no index is written.
.PP
.B level=<-1|0|1|9|11>:
set compression level of the output file if bgzf=1. Valid
values are
.IP -1:
zlib/gzip default compression level
.IP 0:
uncompressed
.IP 1:
zlib/gzip level 1 (fast) compression
.IP 9:
zlib/gzip level 9 (best) compression
.P
If libmaus has been compiled with support for igzip (see
https://software.intel.com/en-us/articles/igzip-a-high-performance-deflate-compressor-with-optimizations-for-genomic-data)
then an additional valid value is
.IP 11:
igzip compression
.PP
.B minlength=<[0]>
Minimum length. Reads shorter than this will be discarded. By default all
reads are kept.
.SH AUTHOR
Written by German Tischler.
.SH "REPORTING BUGS"
Report bugs to <tischler@mpi-cbg.de>
.SH COPYRIGHT
Copyright \(co 2009-2014 German Tischler, \(co 2011-2014 Genome Research Limited.
License GPLv3+: GNU GPL version 3 <http://gnu.org/licenses/gpl.html>
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.


debian/changelog
biobambam2 (2.0.80) unstable; urgency=medium

  * add minlength parameter in normalisefasta

 -- German Tischler <tischler@mpi-cbg.de>  Tue, 24 Oct 2017 13:21:19 +0200

biobambam2 (2.0.79) unstable; urgency=medium

  * update to libmaus2 2.0.398

 -- German Tischler <tischler@mpi-cbg.de>  Fri, 06 Oct 2017 11:37:27 +0200

biobambam2 (2.0.78) unstable; urgency=medium

  * update libmaus2 requirement

 -- German Tischler <tischler@mpi-cbg.de>  Wed, 27 Sep 2017 10:56:56 +0200

biobambam2 (2.0.77) unstable; urgency=medium

  * add sorting by string aux field in bamsort

 -- German Tischler <tischler@mpi-cbg.de>  Wed, 27 Sep 2017 10:38:19 +0200

biobambam2 (2.0.76) unstable; urgency=medium

  * add rcsupport switch in bamsormadup

 -- German Tischler <tischler@mpi-cbg.de>  Tue, 08 Aug 2017 16:49:12 +0200

biobambam2 (2.0.75) unstable; urgency=medium

  * add mc field in bammarkduplicatesopt

 -- German Tischler <tischler@mpi-cbg.de>  Thu, 03 Aug 2017 12:24:54 +0200

biobambam2 (2.0.74) unstable; urgency=medium

  * add bamfillquery

 -- German Tischler <tischler@mpi-cbg.de>  Wed, 02 Aug 2017 10:00:47 +0200

biobambam2 (2.0.73) unstable; urgency=medium

  * add bamfilterrefid
  * allow O key in bammerge (disable check for stdout being a terminal if O key is set)
  * add bamrefinterval program

 -- German Tischler <tischler@mpi-cbg.de>  Tue, 20 Jun 2017 14:48:51 +0200

biobambam2 (2.0.72) unstable; urgency=medium

  * add output file prefix for default read group in bamtofastq

 -- German Tischler <tischler@mpi-cbg.de>  Thu, 16 Mar 2017 10:22:23 +0100

biobambam2 (2.0.71) unstable; urgency=medium

  * add outputperreadgrouprgsm and outputperreadgroupprefix options in bamtofastq

 -- German Tischler <tischler@mpi-cbg.de>  Thu, 16 Mar 2017 10:06:40 +0100

biobambam2 (2.0.70) unstable; urgency=medium

  * add bamscrapcount

 -- German Tischler <tischler@mpi-cbg.de>  Tue, 07 Mar 2017 17:05:10 +0100

biobambam2 (2.0.69) unstable; urgency=medium

  * encode ref id of first read in file in bamexplode file names

 -- German Tischler <tischler@mpi-cbg.de>  Fri, 27 Jan 2017 13:32:37 +0100

biobambam2 (2.0.68) unstable; urgency=medium

  * added HD line header merging in bamcat

 -- German Tischler <tischler@mpi-cbg.de>  Fri, 27 Jan 2017 11:33:22 +0100

biobambam2 (2.0.67) unstable; urgency=medium

  * no adapter clipping if remaining sequence would be a single base (fix by Andrew Whitwham)
  * add bamauxmerge

 -- German Tischler <tischler@mpi-cbg.de>  Thu, 26 Jan 2017 17:30:44 +0100

biobambam2 (2.0.66) unstable; urgency=medium

  * add parserbuffersize parameter in bamsormadup 
  * add optminpixeldif parameter passing in bamsormadup
  * add bamfilterlength program

 -- German Tischler <tischler@mpi-cbg.de>  Thu, 05 Jan 2017 12:06:53 +0100

biobambam2 (2.0.65) unstable; urgency=medium

  * fix bug in bammarkduplicatesopt where last set of duplicates was no marked with od fields

 -- German Tischler <tischler@mpi-cbg.de>  Wed, 30 Nov 2016 12:14:14 +0100

biobambam2 (2.0.64) unstable; urgency=medium

  * add addmatecigar option in bammarkduplicatesopt

 -- German Tischler <tischler@mpi-cbg.de>  Tue, 29 Nov 2016 12:31:07 +0100

biobambam2 (2.0.63) unstable; urgency=medium

  * fix bug in bammarkduplicates/bammarkduplicates2 concerning input via pipe, markthreads>1 and rewritebam=2 

 -- German Tischler <tischler@mpi-cbg.de>  Mon, 28 Nov 2016 10:47:09 +0100

biobambam2 (2.0.62) unstable; urgency=medium

  * fix bin computation for unmapped reads in bamsormadup with SAM input 

 -- German Tischler <tischler@mpi-cbg.de>  Fri, 25 Nov 2016 13:34:04 +0100

biobambam2 (2.0.61) unstable; urgency=medium

  * filter out broken alignments (cigar reaching past end of reference) in bamrecalculatecigar instead of quitting

 -- German Tischler <tischler@mpi-cbg.de>  Mon, 21 Nov 2016 09:06:09 +0100

biobambam2 (2.0.60) unstable; urgency=medium

  * add bamnumericalindex program

 -- German Tischler <tischler@mpi-cbg.de>  Thu, 17 Nov 2016 15:13:15 +0100

biobambam2 (2.0.59) unstable; urgency=medium

  * adapt bamparsort to new libmaus2 API
  * add numeric index construction in bamindex

 -- German Tischler <tischler@mpi-cbg.de>  Mon, 14 Nov 2016 14:09:29 +0100

biobambam2 (2.0.58) unstable; urgency=medium

  * Update to libmaus2 version 2.0.268 (fixes for mate cigar string copying in bamsormadup)

 -- German Tischler <tischler@mpi-cbg.de>  Tue, 01 Nov 2016 10:35:57 +0100

biobambam2 (2.0.57) unstable; urgency=medium

  * Added an explicit iRODS disconnect to avoid the iRODS 4.19 crash. (patch by Andrew Whitwham)

 -- German Tischler <tischler@mpi-cbg.de>  Sun, 18 Sep 2016 18:56:06 +0200

biobambam2 (2.0.56) unstable; urgency=medium

  * fix compilation after API change in libmaus2

 -- German Tischler <tischler@mpi-cbg.de>  Thu, 25 Aug 2016 15:51:00 +0200

biobambam2 (2.0.55) unstable; urgency=medium

  * conditional compilation for PTHREAD_MUTEX_RECURSIVE/PTHREAD_MUTEX_RECURSIVE_NP

 -- German Tischler <tischler@mpi-cbg.de>  Wed, 03 Aug 2016 12:12:34 +0200

biobambam2 (2.0.54) unstable; urgency=medium

  * implement the tags option in bamtofastq (support copying of aux fields from SAM/BAM/CRAM to FastQ)

 -- German Tischler <tischler@mpi-cbg.de>  Tue, 02 Aug 2016 16:25:48 +0200

biobambam2 (2.0.53) unstable; urgency=medium

  * update to libmaus2 version 2.0.240
  * merge/import Andrew Whitwham's parallel adapter finding code

 -- German Tischler <tischler@mpi-cbg.de>  Mon, 01 Aug 2016 10:41:37 +0200

biobambam2 (2.0.52) unstable; urgency=medium

  * add bamfiltermc program

 -- German Tischler <tischler@mpi-cbg.de>  Fri, 15 Jul 2016 12:39:39 +0200

biobambam2 (2.0.51) unstable; urgency=medium

  * only copy mate cigar string if mate is mapped in bamsort

 -- German Tischler <tischler@mpi-cbg.de>  Fri, 15 Jul 2016 10:43:26 +0200

biobambam2 (2.0.50) unstable; urgency=medium

  * do not copy CIGAR data of unmapped mates in bamsormadup 

 -- German Tischler <tischler@mpi-cbg.de>  Tue, 05 Jul 2016 16:13:07 +0200


src/programs/normalisefasta.cpp
/**
    biobambam
    Copyright (C) 2009-2014 German Tischler
    Copyright (C) 2011-2014 Genome Research Limited

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
**/
#include <biobambam2/BamBamConfig.hpp>
#include <biobambam2/Licensing.hpp>

#include <config.h>

#include <libmaus2/bambam/BamBlockWriterBaseFactory.hpp>
#include <libmaus2/fastx/BgzfFastAIndexEntry.hpp>
#include <libmaus2/fastx/FastABgzfIndex.hpp>
#include <libmaus2/fastx/StreamFastAReader.hpp>
#include <libmaus2/lz/BgzfDeflate.hpp>
#include <libmaus2/util/ArgInfo.hpp>
#include <libmaus2/util/MemUsage.hpp>


static unsigned int getDefaultCols()
{
	return 80;
}

static unsigned int getDefaultNRandom()
{
	return 0;
}

static unsigned int getDefaultToUpper()
{
	return 0;
}

static unsigned int getDefaultBgzf()
{
	return 0;
}

static int getDefaultLevel()
{
	return -1;
}

std::string stripName(std::string const & s)
{
	uint64_t j = 0;
	while ( j < s.size() && isspace(s[j]) )
		++j;

	uint64_t i = j;
	while ( i < s.size() && !isspace(s[i]) )
		++i;

	return s.substr(j,i-j);
}

static unsigned int getDefaultMinLength()
{
	return 0;
}

void normalisefastaUncompressed(libmaus2::util::ArgInfo const & arginfo)
{
	libmaus2::fastx::StreamFastAReaderWrapper in(std::cin);
	libmaus2::fastx::StreamFastAReaderWrapper::pattern_type pattern;
	unsigned int const cols = arginfo.getValue<unsigned int>("cols",getDefaultCols());
	unsigned int const nrandom = arginfo.getValue<unsigned int>("nrandom",getDefaultNRandom());
	unsigned int const ctoupper = arginfo.getValue<unsigned int>("toupper",getDefaultToUpper());
	unsigned int const minlength = arginfo.getValue<unsigned int>("minlength",getDefaultMinLength());
	libmaus2::random::Random::setup(::time(0));
	uint64_t offset = 0;

	while ( in.getNextPatternUnlocked(pattern) )
	{
		std::string const name = pattern.getStringId();
		std::string const shortname = stripName(name);

		if ( nrandom )
		{
			std::string & s = pattern.spattern;
			for ( uint64_t i = 0; i < s.size(); ++i )
				if ( libmaus2::fastx::mapChar(s[i]) > 3 )
					s[i] = libmaus2::fastx::remapChar(libmaus2::random::Random::rand8()&3);
			pattern.pattern = pattern.spattern.c_str();
		}

		if ( ctoupper )
		{
			std::string & s = pattern.spattern;
			for ( uint64_t i = 0; i < s.size(); ++i )
				s[i] = ::toupper(s[i]);
			pattern.pattern = pattern.spattern.c_str();
		}

		if ( pattern.spattern.size() < minlength )
			continue;

		std::cerr <<  shortname << "\t" << pattern.patlen << "\t"
			<< offset+pattern.getStringId().size()+2 << "\t" << cols << "\t" << cols+1 << std::endl;

		pattern.printMultiLine(std::cout,cols,offset);
	}

	std::cout << std::flush;
}

void normalisefastaBgzf(libmaus2::util::ArgInfo const & arginfo, std::ostream & out)
{
	libmaus2::fastx::StreamFastAReaderWrapper in(std::cin);
	libmaus2::fastx::StreamFastAReaderWrapper::pattern_type pattern;
	int const level = libmaus2::bambam::BamBlockWriterBaseFactory::checkCompressionLevel(arginfo.getValue("level",getDefaultLevel()));
	std::string const indexfn = arginfo.getUnparsedValue("index","");
	unsigned int const minlength = arginfo.getValue<unsigned int>("minlength",getDefaultMinLength());

	::libmaus2::bambam::BamBlockWriterBaseFactory::checkCompressionLevel(level);

	libmaus2::lz::BgzfDeflate<std::ostream> defl(out,level,false /* full flush */);
	uint64_t const inbufsize = defl.getInputBufferSize();
	uint64_t zoffset = 0;
	uint64_t ioffset = 0;
	std::vector<libmaus2::fastx::BgzfFastAIndexEntry> index;
	std::ostringstream indexstr;

	ioffset += libmaus2::util::NumberSerialisation::serialiseNumber(indexstr,inbufsize);
	uint64_t patid = 0;

	while ( in.getNextPatternUnlocked(pattern) )
	{
		std::string const name = pattern.getStringId();
		std::string const shortname = stripName(name);
		std::string const & spat = pattern.spattern;
		char const * cpat = spat.c_str();
		uint64_t const patlen = spat.size();
		uint64_t const numblocks = (patlen + inbufsize - 1)/inbufsize;

		if ( patlen < minlength )
			continue;

		index.push_back(libmaus2::fastx::BgzfFastAIndexEntry(shortname,patid++,ioffset));

		ioffset += libmaus2::util::StringSerialisation::serialiseString(indexstr,name);
		ioffset += libmaus2::util::StringSerialisation::serialiseString(indexstr,shortname);
		ioffset += libmaus2::util::NumberSerialisation::serialiseNumber(indexstr,patlen);
		ioffset += libmaus2::util::NumberSerialisation::serialiseNumber(indexstr,zoffset);
		ioffset += libmaus2::util::NumberSerialisation::serialiseNumber(indexstr,numblocks);

		std::ostringstream nameostr;
		nameostr << '>' << name << '\n';
		std::string const nameser = nameostr.str();

		std::pair<uint64_t,uint64_t> const P0 = defl.writeSyncedCount(nameser.c_str(),nameser.size());
		zoffset += P0.second;

		uint64_t o = 0;
		while ( o != patlen )
		{
			assert ( o % inbufsize == 0 );
			uint64_t const towrite = std::min(patlen-o,inbufsize);
			std::pair<uint64_t,uint64_t> const P1 = defl.writeSyncedCount(cpat,towrite);

			ioffset += libmaus2::util::NumberSerialisation::serialiseNumber(indexstr,zoffset);

			zoffset += P1.second;
			o += towrite;
			cpat += towrite;
		}

		ioffset += libmaus2::util::NumberSerialisation::serialiseNumber(indexstr,zoffset);

		std::pair<uint64_t,uint64_t> const Pn = defl.writeSyncedCount("\n",1);
		zoffset += Pn.second;
	}

	defl.flush();
	out << std::flush;

	uint64_t const imetaoffset = ioffset;

	ioffset += libmaus2::util::NumberSerialisation::serialiseNumber(indexstr,index.size());
	for ( uint64_t i = 0; i < index.size(); ++i )
		ioffset += libmaus2::util::NumberSerialisation::serialiseNumber(indexstr,index[i].ioffset);

	libmaus2::util::NumberSerialisation::serialiseNumber(indexstr,imetaoffset);

	if ( indexfn.size() )
	{
		std::string const & sindex = indexstr.str();
		libmaus2::aio::OutputStreamInstance indexCOS(indexfn);
		indexCOS.write(sindex.c_str(),sindex.size());
		indexCOS.flush();
	}
}

void normalisefasta(libmaus2::util::ArgInfo const & arginfo)
{
	if ( arginfo.getValue<unsigned int>("bgzf",getDefaultBgzf()) )
		normalisefastaBgzf(arginfo,std::cout);
	else
		normalisefastaUncompressed(arginfo);
}

int main(int argc, char * argv[])
{
	try
	{
		libmaus2::timing::RealTimeClock rtc; rtc.start();

		::libmaus2::util::ArgInfo const arginfo(argc,argv);

		for ( uint64_t i = 0; i < arginfo.restargs.size(); ++i )
			if (
				arginfo.restargs[i] == "-v"
				||
				arginfo.restargs[i] == "--version"
			)
			{
				std::cerr << ::biobambam2::Licensing::license();
				return EXIT_SUCCESS;
			}
			else if (
				arginfo.restargs[i] == "-h"
				||
				arginfo.restargs[i] == "--help"
			)
			{
				std::cerr << ::biobambam2::Licensing::license() << std::endl;
				std::cerr << "Key=Value pairs:" << std::endl;
				std::cerr << std::endl;

				std::vector< std::pair<std::string,std::string> > V;

				V.push_back ( std::pair<std::string,std::string> ( std::string("cols=<[")+libmaus2::util::NumberSerialisation::formatNumber(getDefaultCols(),0)+"]>", "column width" ) );
				V.push_back ( std::pair<std::string,std::string> ( std::string("nrandom=<[")+libmaus2::util::NumberSerialisation::formatNumber(getDefaultNRandom(),0)+"]>", "replace N bases randomly by A,C,G,T" ) );
				V.push_back ( std::pair<std::string,std::string> ( std::string("toupper=<[")+libmaus2::util::NumberSerialisation::formatNumber(getDefaultToUpper(),0)+"]>", "convert bases to upper case" ) );
				V.push_back ( std::pair<std::string,std::string> ( std::string("bgzf=<[")+libmaus2::util::NumberSerialisation::formatNumber(getDefaultBgzf(),0)+"]>", "compress output" ) );
				V.push_back ( std::pair<std::string,std::string> ( std::string("index=<>"), "file name for index if bgzf=1 (no index is created if key is not given)" ) );
				V.push_back ( std::pair<std::string,std::string> ( std::string("level=<[")+::biobambam2::Licensing::formatNumber(getDefaultLevel())+"]>", std::string("compression level if bgzf=1 (") + libmaus2::bambam::BamBlockWriterBaseFactory::getLevelHelpText() + std::string(")") ) );
				V.push_back ( std::pair<std::string,std::string> ( std::string("minlength=<[")+::biobambam2::Licensing::formatNumber(getDefaultMinLength())+"]>", std::string("minimum length") ) );

				::biobambam2::Licensing::printMap(std::cerr,V);

				std::cerr << std::endl;
				return EXIT_SUCCESS;
			}

		normalisefasta(arginfo);
	}
	catch(std::exception const & ex)
	{
		std::cerr << ex.what() << std::endl;
		return EXIT_FAILURE;
	}

}
